# Day 12: Subterranean Sustainability
[link](https://adventofcode.com/2018/day/12)

## Part 1
After `20` generations, what is the sum of the numbers of all pots which contain a plant?

### Inputs

In [1]:
puzzle_in = """initial state: ##....#.#.#...#.#..#.#####.#.#.##.#.#.#######...#.##....#..##....#.#..##.####.#..........#..#...#

..#.# => #
.#### => #
#.... => .
####. => #
...## => .
.#.#. => .
..#.. => .
##.#. => .
#.#.# => #
..... => .
#.#.. => .
....# => .
.#..# => .
###.# => #
#..#. => .
##### => .
...#. => #
#.##. => #
.#.## => #
#..## => #
.##.. => #
##.## => .
..### => .
###.. => .
##..# => #
.#... => #
.###. => #
#.### => .
.##.# => .
#...# => #
##... => .
..##. => .
"""

In [2]:
test_in = """initial state: #..#.#..##......###...###

...## => #
..#.. => #
.#... => #
.#.#. => #
.#.## => #
.##.. => #
.#### => #
#.#.# => #
#.### => #
##.#. => #
##.## => #
###.. => #
###.# => #
####. => #
"""

### Logic

In [3]:
class Machine:
  generation = 0

  def __init__(self, input_str):
    self.rules = set()
    for line in input_str.split('\n'):
      if '=>' in line:
        pots, outcome = line.split('=>')
        if outcome.strip() == '#':
          rule = [c for c in pots.strip() if c in '#.']
          assert len(rule)==5
          self.rules.add(''.join(rule))
      elif line.startswith('initial state:'):
        self.state = ''.join(c for c in line.split(':')[1] if c in '#.')
        self.state_offset = 0

    assert '#' in self.state
    assert len(self.rules)>1

  def evolve(self):
    filler = '....'
    state = filler + self.state + filler
    new_state = []
    for i in range(2, len(state)-2):
      batch = state[i-2:i+3]
      new_state.append('#' if batch in self.rules else '.')

    self.state = ''.join(new_state)
    self.state_offset += 2

    offset_change = 0
    for i in range(5):
      if self.state[i] == '#':
        break;
      offset_change+=1
    self.state_offset -= offset_change
    self.state = self.state[offset_change:].strip('.')
    self.generation += 1

  def __repr__(self):
    return f'{self.generation:02} {self.state}'

  def hash_code(self):
    return sum(i-self.state_offset for i,c in enumerate(self.state) if c=='#')

In [4]:
test_machine = Machine(test_in)
print(test_machine)

for i in range(20):
  test_machine.evolve()
  print(test_machine)

assert test_machine.state.strip('.')=='#....##....#####...#######....#.#..##'
assert test_machine.hash_code() == 325, test_machine.hash_code()

00 #..#.#..##......###...###
01 #...#....#.....#..#..#..#
02 ##..##...##....#..#..#..##
03 #.#...#..#.#....#..#..#...#
04 #.#..#...#.#...#..#..##..##
05 #...##...#.#..#..#...#...#
06 ##.#.#....#...#..##..##..##
07 #..###.#...##..#...#...#...#
08 #....##.#.#.#..##..##..##..##
09 ##..#..#####....#...#...#...#
10 #.#..#...#.##....##..##..##..##
11 #...##...#.#...#.#...#...#...#
12 ##.#.#....#.#...#.#..##..##..##
13 #..###.#....#.#...#....#...#...#
14 #....##.#....#.#..##...##..##..##
15 ##..#..#.#....#....#..#.#...#...#
16 #.#..#...#.#...##...#...#.#..##..##
17 #...##...#.#.#.#...##...#....#...#
18 ##.#.#....#####.#.#.#...##...##..##
19 #..###.#..#.#.#######.#.#.#..#.#...#
20 #....##....#####...#######....#.#..##


In [5]:
puzzle_machine = Machine(puzzle_in)

for i in range(20):
  puzzle_machine.evolve()
  print(f'{puzzle_machine.hash_code():05} {puzzle_machine}')

assert puzzle_machine.hash_code()==2349

print(puzzle_machine.hash_code())

01919 01 #...##.#..####....##.#.##.#.###..#.##.#...#..######...#..#.#...##...#....###..#........#....###.#
01849 02 #.##......#.##........###..##.#.#.###...###..#.#..#..##...#..##..#.##.#....#.#..#......#.#....##..#
01887 03 #####.....#####.........#.##...#.##.#..#..#.#.#......#.#.##..#.##.###...#..##.....#....##..#....##..#
02064 04 #.#.......#.#.........#####.#####........#.#..#....##.#####.###...#..##..#.#....#.#....##..#....##..#
01821 05 ##..#.....##..#.........#.##..#.#........##.....#.......#.##..#..##..#.##.#..#..##..#....##..#....##..#
02017 06 ##..#.....##..#.......######.#..#........#....#.#.....######...#.##.###.......#.##..#....##..#....##..#
02013 07 ##..#.....##..#.......#..##.....#......#.#..##..#.....#..#..#####...#.......######..#....##..#....##..#
01814 08 ##..#.....##..#.....#..#.#....#.#....##...#.##..#...#.....#.#.#..##.#.......#..#.#..#....##..#....##..#
01948 09 ##..#.....##..#...#...#..#..##..#....#.#######..###.#...##.#...#....#.....#...#.....#....##..#....##.

**Incorrect answers:**

- `2022` is too low

**Correct answer:** `2349`

In [6]:
puzzle_machine = Machine(puzzle_in)

for i in range(200):
  puzzle_machine.evolve()

previous_hash = puzzle_machine.hash_code()

assert previous_hash == 9274

for i in range(50000):
  puzzle_machine.evolve()
  hash_code = puzzle_machine.hash_code()
  assert hash_code-previous_hash == 42
  previous_hash = hash_code

At step # `200`, the hash code is `9274`. It increases by 42 on each following step. Therefore, the hash code after step `50000000000` would be:

In [9]:
9274 + 42*(50000000000-200)

2100000000874

**Incorrect answers**

- `2100000000874` is too low
- `2100000000916` is too low